internal/cmd/cloud/web_flow.go

```diff
@@ -0,0 +1,85 @@
+package cloud
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"net/url"
+)
+
+// Flow holds the state for the steps of our OAuth-like Web Application flow.
+type Flow struct {
+	server *localServer
+	state  string
+}
+
+// InitFlow creates a new Flow instance by detecting a locally available port number.
+func InitFlow() (*Flow, error) {
+	server, err := bindLocalServer()
+	if err != nil {
+		return nil, err
+	}
+
+	state := randomString(20)
+
+	return &Flow{
+		server: server,
+		state:  state,
+	}, nil
+}
+
+// BrowserURL appends GET query parameters to baseURL and returns the url that the user should
+// navigate to in their web browser.
+func (f *Flow) BrowserURL(baseURL string) (string, error) {
+	const callbackURL = "http://localhost:8080/cloud/auth/callback"
+
+	ru, err := url.Parse(callbackURL)
+	if err != nil {
+		return "", err
+	}
+
+	ru.Host = fmt.Sprintf("%s:%d", ru.Hostname(), f.server.Port())
+	f.server.CallbackPath = ru.Path
+
+	q := url.Values{}
+	q.Set("redirect_url", ru.String())
+	q.Set("state", f.state)
+
+	return fmt.Sprintf("%s?%s", baseURL, q.Encode()), nil
```

Comment:
I think maybe we should set callback path on the local server itself (in `bindLocalServer`), and not set it via the flow after calling BrowserURL(). As that just makes this server and flow thing stateful and error prone (if you start server before getting browser URL then there is no callback path on the server).

Then we can simplify this call to just be a deterministic URL formatting operation:

```suggestion
	u, err := url.Parse(baseURL)
	if err != nil {
		return "", err
	}
	
	q := u.Query()
	q.Set("redirect_url", fmt.Sprintf("http://localhost:%d%s", f.server.Port(), f.server.CallbackPath))
	q.Set("state", f.state)
	u.RawValues = q.Encode()

	return u.String, nil
```

Reply:
This race does actually exist too, because we're calling BrowserURL concurrently with StartServer. So there is a small opportunity for a request to come in and be handled when the path is not set.